mergify_engine/squash_pull.py

```diff
@@ -0,0 +1,160 @@
+        await git("fetch", "--quiet", "origin", head_branch)
+
+        await git("remote", "add", "upstream", f"{config.GITHUB_URL}/{base_repo}")
+        await git("config", "remote.upstream.promisor", "true")
+        await git("config", "remote.upstream.partialclonefilter", "blob:none")
+        await git("fetch", "--quiet", "upstream", base_branch)
+
+        await git("checkout", "-q", "-b", tmp_branch, f"upstream/{base_branch}")
+
+        await git("merge", "--squash", "--no-edit", f"origin/{head_branch}")
+        await git("commit", "-m", squash_message)
+
+        await git(
+            "push",
+            "--verbose",
+            "origin",
+            f"{tmp_branch}:{head_branch}",
+            "--force-with-lease",
+        )
+
+        expected_sha = (await git("log", "-1", "--format=%H")).strip()
+        # NOTE(sileht): We store this for dismissal action
+        # FIXME(sileht): use a more generic name for the key
+        await ctxt.redis.setex(f"branch-update-{expected_sha}", 60 * 60, expected_sha)
+    except gitter.GitAuthenticationFailure:
+        raise
+    except gitter.GitErrorRetriable as e:
+        raise SquashNeedRetry(
+            f"Git reported the following error:\n```\n{e.output}\n```\n"
+        )
+    except gitter.GitFatalError as e:
+        raise SquashFailure(
+            f"Git reported the following error:\n```\n{e.output}\n```\n"
+        )
+    except gitter.GitError as e:
+        ctxt.log.error(
+            "squash failed",
+            output=e.output,
+            returncode=e.returncode,
+            exc_info=True,
+        )
+        raise SquashFailure("")
+    except Exception:  # pragma: no cover
+        ctxt.log.error("squash failed", exc_info=True)
+        raise SquashFailure("")
+    finally:
+        await git.cleanup()
+
+
+async def squash(
+    ctxt: context.Context,
+    message: str,
+    bot_account: typing.Optional[str] = None,
+) -> None:
+
+    if ctxt.pull["commits"] <= 1:
+        return
+
+    try:
+        users = await user_tokens.UserTokens.select_users_for(ctxt, bot_account)
+    except user_tokens.UserTokensUserNotFound:
+        raise SquashFailure(
+            f"Unable to squash: user `{bot_account}` is unknown. "
+            f"Please make sure `{bot_account}` has logged in Mergify dashboard."
+        )
+
+    for user in users:
+        try:
+            await _do_squash(ctxt, user, message)
+        except gitter.GitAuthenticationFailure as e:
+            ctxt.log.info(
+                "authentification failure, will retry another token: %s",
+                e,
+                login=user["login"],
+            )
+        else:
+            return
+
+    ctxt.log.warning("unable to squash pull request: no tokens are valid")
+
+    if ctxt.pull_from_fork and ctxt.pull["base"]["repo"]["private"]:
+        raise SquashFailure(
+            "Squashing a branch for a forked private repository is not supported by GitHub"
+        )
```

Comment:
Why is that so late in the process? We should do this first thing to bypass the retry loop?

Reply:
When we wrote this for branch_updater, we was not 100% confident about the conditions for this to work or not.
That why we wrote the message only after we try all tokens.

Reply:
Can we check the log in prod to see if that happen with the branch updater?

Reply:
I don't think we can catch that in logs.